File: frontend/src/microservices/feedback/api/OpenComplaintManagerThreadPlanner.ts
// 本文件定义 feedback 模块的 `OpenComplaintManagerThreadPlanner` 接口入口，负责把投诉消息打开为站内客服可处理的管理线程。
import type { OpenComplaintManagerThreadPlannerRequest } from '@/microservices/feedback/objects/OpenComplaintManagerThreadPlannerRequest'
import type { FeedbackThreadDetailsPlannerResponse } from '@/microservices/feedback/objects/FeedbackThreadDetailsPlannerResponse'
import { executeJsonApiRequest } from '@/shared-kernel/api/ApiTransport'

type BackendFeedbackThreadDetailsPlannerResponse = {
  thread: Omit<FeedbackThreadDetailsPlannerResponse, 'messages'>
  messages: FeedbackThreadDetailsPlannerResponse['messages']
  managerActorLogoAssetPath?: string | null
  siteAdminActorLogoAssetPath?: string | null
}

const flattenThread = (response: BackendFeedbackThreadDetailsPlannerResponse): FeedbackThreadDetailsPlannerResponse => ({
  ...response.thread,
  managerActorLogoAssetPath: response.managerActorLogoAssetPath ?? response.thread.managerActorLogoAssetPath ?? null,
  siteAdminActorLogoAssetPath: response.siteAdminActorLogoAssetPath ?? response.thread.siteAdminActorLogoAssetPath ?? null,
  messages: response.messages,
})

export const openComplaintManagerThread = (payload: OpenComplaintManagerThreadPlannerRequest): Promise<FeedbackThreadDetailsPlannerResponse> =>
  executeJsonApiRequest<BackendFeedbackThreadDetailsPlannerResponse>('/OpenComplaintManagerThreadPlanner', 'POST', payload).then(flattenThread)

File: backend/src/main/scala/microservices/feedback/api/OpenComplaintManagerThreadPlanner.scala
// 本文件提供 content 域的投诉管理会话打开入口。
package com.typesafe.travel.feedback.domain

import cats.effect.IO
import com.typesafe.travel.api.routes.ConnectionApiPlan
import com.typesafe.travel.feedback.domain.*
import com.typesafe.travel.persistence.feedback.FeedbackPlannerPlainSql
import com.typesafe.travel.shared.kernel.*

import java.sql.Connection

object OpenComplaintManagerThreadPlanner extends ConnectionApiPlan[OpenComplaintManagerThreadPlannerRequest, FeedbackThreadDetailsPlannerResponse]:
  override val name: String = "OpenComplaintManagerThreadPlanner"
  override def plan(input: OpenComplaintManagerThreadPlannerRequest, connection: Connection): IO[FeedbackThreadDetailsPlannerResponse] =
    val messageId = SupportMessageId(input.complaintMessageId)
    for
      siteThreads <- FeedbackPlannerPlainSql.listSiteAdminParticipantThreads(connection, "user", input.siteAdminActorId)
      cardMessage <- siteThreads
        .map(thread => FeedbackPlannerPlainSql.findMessage(connection, thread.threadId, messageId))
        .foldLeft(IO.pure(Option.empty[FeedbackMessage])) { (current, next) =>
          current.flatMap {
            case found @ Some(_) => IO.pure(found)
            case None            => next
          }
        }
      payload <- cardMessage.flatMap(_.complaintPayload) match
        case Some(found) => IO.pure(found)
        case None        => IO.raiseError(new IllegalArgumentException(s"Complaint card '${input.complaintMessageId}' was not found"))
      managerThreadId <- payload.managerThreadId match
        case Some(threadId) => IO.pure(SupportTicketId(threadId))
        case None           => IO.raiseError(new IllegalArgumentException("Complaint card does not have a manager thread"))
      managerThread <- requireFeedbackThread(connection, managerThreadId)
      response <- toThreadDetailsResponse(connection, managerThread)
    yield response

